nomos-core/chain-defs/src/proofs/leader_proof.rs

```diff
@@ -1,100 +1,257 @@
-        private_inputs: &LeaderPrivate,
-        prover: &dyn risc0_zkvm::Prover,
-    ) -> Result<Self, Error> {
-        let env = risc0_zkvm::ExecutorEnv::builder()
-            .write(&public_inputs)
-            .unwrap()
-            .write(&private_inputs)
-            .unwrap()
-            .build()
-            .unwrap();
-
+impl Groth16LeaderProof {
+    pub fn prove(private: &LeaderPrivate, voucher_cm: VoucherCm) -> Result<Self, Error> {
         let start_t = std::time::Instant::now();
+        let (proof, entropy_contribution) = if std::env::var(POL_PROOF_DEV_MODE).is_ok() {
+            tracing::warn!(
+                "Proofs are being generated in dev mode. This should never be used in production."
+            );
+            (
+                groth16::CompressedGroth16Proof::new(
+                    GenericArray::default(),
+                    GenericArray::default(),
+                    GenericArray::default(),
+                ),
+                Fr::ZERO,
+            )
+        } else {
+            let (proof, verif_inputs) =
+                pol::prove(&private.input).map_err(Error::PoLProofFailed)?;
+            (proof, verif_inputs.entropy_contribution.into_inner())
+        };
+        tracing::debug!("groth16 prover time: {:.2?}", start_t.elapsed(),);
+
+        let leader_key = private.pk;
 
-        // ATTENTION: producing a groth16 proof currently requires x86 with docker
-        // support
-        let opts = risc0_zkvm::ProverOpts::groth16();
-        let prove_info =
-            prover.prove_with_opts(env, nomos_risc0_proofs::PROOF_OF_LEADERSHIP_ELF, &opts)?;
-
-        tracing::debug!(
-            "STARK prover time: {:.2?}, total_cycles: {}",
-            start_t.elapsed(),
-            prove_info.stats.total_cycles
-        );
-        // extract the receipt.
         Ok(Self {
-            public_inputs,
-            risc0_receipt: prove_info.receipt,
+            proof,
+            entropy_contribution,
+            leader_key,
+            voucher_cm,
         })
     }
+
+    #[must_use]
+    pub const fn proof(&self) -> &pol::PoLProof {
+        &self.proof
+    }
 }
 
 pub trait LeaderProof {
     /// Verify the proof against the public inputs.
     fn verify(&self, public_inputs: &LeaderPublic) -> bool;
 
     /// Get the entropy used in the proof.
-    fn entropy(&self) -> [u8; 32];
+    fn entropy(&self) -> Fr;
+
+    fn leader_key(&self) -> &ed25519_dalek::VerifyingKey;
+
+    fn voucher_cm(&self) -> &VoucherCm;
 }
 
-impl LeaderProof for Risc0LeaderProof {
+impl LeaderProof for Groth16LeaderProof {
     fn verify(&self, public_inputs: &LeaderPublic) -> bool {
-        // The risc0 proof is valid by contract
-        public_inputs == &self.public_inputs
+        if std::env::var(POL_PROOF_DEV_MODE).is_ok() {
+            tracing::warn!(
+                "Proofs are being verified in dev mode. This should never be used in production."
+            );
+            return true;
```

Comment:
My understanding of DEV_MODE in risc0 was that it was effectively providing the witness as the proof. Then a proof passing verification in dev mode implies the proof is valid in production mode as well.

I think the same would be useful for us with Groth16 as well. Especially for tests where we do want to test PoL _behaviour_ without incurring the proving costs.

But this is something we can do later. Perhaps once we integrate with the wallet, I'll take a look at what would be involved in mirroring the PoL circom proof in rust. (maybe we can even find a circom interperter in rust that will actually validate the PoL constraints directly)